### ISSUE TITLE
✨ Support --mount=type=ssh with dockerBuild()

### ISSUE THREAD

**Issue by chiting** [dagger/dagger#6736]
### What are you trying to do?

I'm trying to use `dockerBuild()` with a Dockerfile containing the statement `RUN --mount=type=ssh ...` to forward the SSH agent to the build container so I can fetch packages from private repositories.

### Why is this important to you?

We can already pass the ssh agent socket when using a container so I thought this feature would be great to close the gap in feature parity when it comes down to build an image with a Dockerfile in dagger.

### How are you currently working around this?

The only workaround I can think about is to use `container().build()` while passing the ssh agent socket but we'd lose the ability to use a Dockerfile.

**Comment by femiagbabiaka** [comment]
Fixing issues like these helps by giving an onramp to the dagger-curious. Instead of having to rewrite services, one can use the code that they have. It also provides a lot of interoperability with other tooling that may rely on physical Dockerfiles.

**Comment by shykes** [comment]
Commenting on behalf of a user on Discord who got stuck in their adoption of Dagger because of this. [Here's a link to their message for reference](https://discord.com/channels/707636530424053791/708371226174685314/1387339749227888732).

**Comment by shykes** [comment]
Is this a gap only in the Dockerfile compat layer, or also in the Dagger API? Is it possible to forward the ssh agent with a native dagger build?

**Comment by marcosnils** [comment]
> Is this a gap only in the Dockerfile compat layer, or also in the Dagger API? Is it possible to forward the ssh agent with a native dagger build?

 yes, this works natively in Dagger. It's currently a gap in the Dockerfile compat layer. We spoke about this and it's not trivial  to implement. @jedevc has more in-detail context given that he's working in a new way which will make this compatibility more robust and complete.

**Comment by sipsma** [comment]
Quick scope-out of what would be required here:
1. buildkit looks for a socket in the current session attachables named `default`, or if the dockerfile includes a `,id=foo` key-value in the mount options, then `foo`
2. dagger already has logic for default use of the caller's SSH socket in our git API ([relevant lines](https://github.com/sipsma/dagger/blob/e71fe8e0e6912a74b47639576cdc5c3570b1a83a/core/schema/git.go#L215-L215))
3. So we'd need to
   * identify any exec-ops in the dockerfile LLB that ask for a socket ([in this walk of that DAG](https://github.com/sipsma/dagger/blob/e71fe8e0e6912a74b47639576cdc5c3570b1a83a/core/container.go#L585-L585))
   * when we solve the dockerfile LLB, hook up the session attachable with the user's default SSH socket and give it the names asked for in the dockerfile (either `default` or value of `id` opt)
4. Also need to be careful around caching here since we don't want to share cache such that you could get cache hits from earlier runs without having any of the auth
   * On the other hand, buildkit doesn't do that at all, that's something we've improved on in the dagger-native API, so I guess you could make an argument we don't need to care about this if we are just going for compat w/ dockerfiles

Worth noting that all the above is still applicable in the future of running buildkitd as a separate service to support dockerfile compat; only difference then is solves will be going to that separate service rather than the internal built-in code we use today.

Overall, wouldn't classify this as utterly trivial, but is mostly plumbing together a few existing components. Customizing the session attachable used for the solve of the dockerfile is potentially the most tedious part.
